File: MobileAPI/iOS/PaymentDemo/ViewController.m
//
//  ViewController.m
//  PaymentTest
//
//  Created by MacRealize on 25/11/14.
//  Copyright (c) 2014 Viva Payments. All rights reserved.
//

#import "ViewController.h"

#if !defined(__has_feature) || ! __has_feature(objc_arc)
#error This Class requires ARC. Either turn on ARC for the project or use -fobjc-arc flag
#endif

// Replace with YOUR Credentials
// See more at https://github.com/VivaPayments/API/wiki/API%20Authentication
static NSString *merchantID = @"90DDA476-CF7C-4CFD-A9CB-23DFE11F131E";
static NSString *apiKey = @"XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX";
static NSString *publicKey = @"u3a1fcKsxynRZwY8zb++1utUYr1vjdGW6okiEX0pJBc=";

@interface ViewController ()

@end

@implementation ViewController


- (void)viewDidLoad {
	[super viewDidLoad];
	// Do any additional setup after loading the view, typically from a nib.
}


- (void)didReceiveMemoryWarning {
	[super didReceiveMemoryWarning];
	// Dispose of any resources that can be recreated.
}


- (IBAction)createOrderPressed:(id)sender
{
	NSNumber *amountInCents = @100;
	
	mobileApi = [MobileAPI newDemoInstance];
	
	[mobileApi setMerchantID:merchantID
					  apiKey:apiKey
				   publicKey:publicKey];
	
	[mobileApi createOrderWithAmount:amountInCents.unsignedLongLongValue
							  params:nil
                  isRecurrentPayment:false
						  completion:^(BOOL success, NSURLResponse *urlResponse, NSDictionary *response, NSError *error)
	 {
		 NSLog(@"Create order response = %@", response);
		 
		 if (success)
		 {
			 // Warning. Order code is a 64-bit unsigned (uint64_t) encapsulated in a NSNumber
			 orderCode = [response objectForKey:@"OrderCode"];
			 [self createCreditCard];
		 }
	 }];
}


- (void) createCreditCard
{
	[mobileApi createCardTokenForCardNumber:@"4111111111111111"
										cvc:@"111"
					   expirationDateString:@"2021-01-01"
							 cardHolderName:@"John Papadopoulos"
								 completion:^(BOOL success, NSURLResponse *urlResponse, NSDictionary *response, NSError *error)
	 {
		 NSLog(@"Create card token response = %@", response);
		 
		 if (success)
		 {
			 cardToken = [response objectForKey:@"Token"];
			 [self checkInstallments];
		 }
	 }];
}


- (void) checkInstallments
{
	[mobileApi checkInstallmentsForCard:@"4111111111111111"
							 completion:^(BOOL success, NSURLResponse *urlResponse, NSDictionary *response, NSError *error)
	 {
		 NSLog(@"Check Installments reponse = %@", response);
		 
		 
		 if (success)
		 {
			 NSInteger MaxInstallments = [[response objectForKey:@"MaxInstallments"] intValue];
			 NSLog(@"Card max installments = %ld", (long) MaxInstallments);

			 [self createTransaction];
		 }
	 }];
}


- (void) createTransaction
{
	[mobileApi createTransactionWithOrderCode:orderCode
								   sourceCode:@"Default"
								 installments:1
                           isRecurrentPayment:false
							  creditCardToken:cardToken
								   completion:^(BOOL success, NSURLResponse *urlResponse, NSDictionary *response, NSError *error)
	 {
		 NSLog(@"Create transaction reponse = %@", response);
		 
		 if (success)
		 {
			 transactionId = [response objectForKey:@"TransactionId"];
			 NSString *StatusId = [response objectForKey:@"StatusId"];
			 NSLog(@"Completed transaction with id %@, status %@", transactionId, StatusId);
		 }
	 }];
}


- (IBAction)createRecurringTransactionPressed:(id)sender
{
    if (transactionId == nil)
        NSLog(@"First create a successfull payment with isRecurrentPayment: true");
    
    NSNumber *amountInCents = @100;
    
    [mobileApi createRecurringTransaction:amountInCents.unsignedLongLongValue
                                   params:nil
                             installments:1
                            transactionID:transactionId //Use transactionId from a previously completed Order that isRecurrentPayment:true
                               completion:^(BOOL success, NSURLResponse *urlResponse, NSDictionary *response, NSError *error)
     {
         NSLog(@"Create Recurring transaction reponse = %@", response);
         
         if (success)
         {
             NSString *transactionId = [response objectForKey:@"TransactionId"];
             NSString *StatusId = [response objectForKey:@"StatusId"];
             NSLog(@"Completed transaction with id %@, status %@", transactionId, StatusId);
         }
     }];
}


@end

File: MobileAPI/ExampleSwift/ExampleSwift/ViewController.swift
//
//  ViewController.swift
//  ExampleSwift
//
//  Created by Mike on 11/30/16.
//  Copyright © 2016 Mike. All rights reserved.
//

import UIKit

class ViewController: UIViewController {
    
    // Change below variables with your merchantID,apikey,publicKey
    var merchantID = "90DDA476-CF7C-4CFD-A9CB-23DFE11F131E"
    var apiKey = "XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX"
    var publicKey = "u3a1fcKsxynRZwY8zb++1utUYr1vjdGW6okiEX0pJBc="
    // If you don't change the above variables you will get [:] as a response
    var cardNum = "4111111111111111"
    var cvv = "111"
    var expdate = "2021-01-01"
    var cardHolder = "John Papadopoulos"
    var cardToken = ""
    var orderCode: NSNumber = 0
    var transactionId = ""
    var mobileApi = MobileAPI.newDemoInstance()
    
    
    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view, typically from a nib.
    }
    
    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
    
    @IBAction func payButton(_ sender: AnyObject) {
        viva()
    }
    
    @IBAction func payRecurringButton(_ sender: Any) {
        createRecurringTransaction()
    }
    
    func viva() {
        
        let amountInCents = UInt64(10037)
        
        mobileApi?.setMerchantID(merchantID, apiKey: apiKey, publicKey: publicKey)
        
        mobileApi?.createOrder(withAmount: amountInCents, params: nil, isRecurrentPayment: false) { (success, urlResponse, response, error) in
            print("Create order response = \(response ?? [:])")
            
            if success {
                self.orderCode = response?["OrderCode"] as? NSNumber ?? 0
                self.createCreditCard()
            }
        }
    }
    
    func createCreditCard() {
        
        mobileApi?.createCardToken(forCardNumber: cardNum, cvc: cvv, expirationDateString: expdate, cardHolderName: cardHolder) { (success, urlResponse, response, error) in
            print("Create card token response = \(response ?? [:])")
            
            if success {
                self.cardToken = response?["Token"] as? String ?? ""
                self.checkInstallments()
                
            }
        }
    }
    
    func checkInstallments() {
        
        mobileApi?.checkInstallments(forCard: cardNum) { (success, urlResponse, response, error) in
            print("Check Installments reponse = \(response ?? [:])")
            
            if success {
                let MaxInstallments = response?["MaxInstallments"] as? NSNumber ?? 0
                print("Card max installments = \(MaxInstallments)")
                
                self.createTransaction()
                
            }
        }
    }
    
    
    
    func createTransaction() {
        
        mobileApi?.createTransaction(withOrderCode: orderCode as NSNumber, sourceCode: "Default", installments: 1, isRecurrentPayment: false, creditCardToken: cardToken) { (success, urlResponse, response, error) in
            print("Create transaction reponse = \(response ?? [:])")
            
            if success {
                
                self.transactionId = response?["TransactionId"] as! String
                let StatusId      = response?["StatusId"]
                print("Completed transaction with id \(self.transactionId), status \(StatusId ?? [:])");
                
            }
        }
    }
    
    func createRecurringTransaction() {
        
        if transactionId == "" {
            print("First create a successfull payment with isRecurrentPayment: true")
            return
        }
        
        let amountInCents = UInt64(10037)
        
        mobileApi?.createRecurringTransaction(amountInCents, params: nil, installments: 1, transactionID: transactionId) { (success, urlResponse, response, error) in
            print("Create transaction reponse = \(response ?? [:])")
            
            if success {
                
                let transactionId = response?["TransactionId"]
                let StatusId      = response?["StatusId"]
                print("Completed transaction with id \(transactionId ?? [:]), status \(StatusId ?? [:])");
                
            }
        }
    }
    
    
}

